src/main/java/org/apache/geronimo/microprofile/impl/jwtauth/jwt/KidMapper.java

```diff
@@ -120,7 +181,44 @@ private String tryLoad(final String value) {
             throw new IllegalArgumentException(e);
         }
 
-        // else direct value
+        // load jwks via url
+        if (jwksUrl != null) {
+            if(reloadJwksRequest != null) {
+                try {
+                    reloadJwksRequest.get();
+                } catch (InterruptedException | ExecutionException e) {
+                    throw new IllegalStateException(e);
+                }
+            }
+                String key = keyMapping.get(value);
+                if (key != null) {
+                    return key;
+                }
+
+        }
         return value;
     }
+
+    private List<JWK> parseKeys(HttpResponse<String> keyResponse) {
+        StringReader stringReader = new StringReader(keyResponse.body());
+        JsonReader jwksReader = readerFactory.createReader(stringReader);
+        JsonObject keySet = jwksReader.readObject();
+        JsonArray keys = keySet.getJsonArray("keys");
+        return keys.stream()
+                .map(JsonValue::asJsonObject)
+                .map(JWK::new)
+                .filter(it -> it.getUse() == null || "sig".equals(it.getUse()))
+                .collect(toList());
+    }
+
+    @PreDestroy
+    private void destroy() {
+        if (backgroundThread != null) {
+            backgroundThread.shutdown();
```

Comment:
+ await (1mn can be hardcoded since it should be almost immediate - one call duration) to try our best to not leak a thread before quitting this method (the class being the thread owner)

side note: the cancel should be done before the await